File: zlzlelzl/leetcode/Array/Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        # 역배열 접근
        # 다음 최대값이 나올 때까지 최소값을 계속 저장
        # 최대값이 갱신되면 최대-최소를 max_profit과 비교

        max_stock, min_stock = -1, -1
        max_profit = 0
        for i in prices[::-1]:

            if max_stock < i:
                max_profit = max(max_profit, max_stock - min_stock)
                max_stock, min_stock = i, i

            min_stock = min(min_stock, i)

            print(max_stock, min_stock)
        max_profit = max(max_profit, max_stock - min_stock)

        return max_profit


prices =\
    [7, 4, 1, 2]
# [2, 4, 1]

s = Solution()
print(s.maxProfit(prices))

File: JunnyJun/Best Time to Buy and Sell Stock.cpp
class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int min = 20000, result = 0, i;

        for (i = 0; i < prices.size(); i++)
        {
            if (prices[i] < min)
            {
                min = prices[i];
            }
            if (result < (prices[i] - min))
            {
                result = (prices[i] - min);
            }
        }
        return result;
    }
};